text/0000-name-resolution.md

```diff
@@ -0,0 +1,649 @@
+* Rust has an implicit prelude - the prelude defines a set of names which are
+  always (unless explicitly opted-out) nameable. The prelude includes macros.
+  Names in the prelude can be shadowed by any other names.
+
+
+# Detailed design
+[design]: #detailed-design
+
+## Guiding principles
+
+We would like the following principles to hold. There may be edge cases where
+they do not, but we would like these to be as small as possible (and prefer they
+don't exist at all).
+
+#### Avoid 'time-travel' ambiguities, or different results of resolution if names
+are resolved in different orders.
+
+Due to macro expansion, it is possible for a name to be resolved and then to
+become ambiguous, or (with rules formulated in a certain way) for a name to be
+resolved, then to be amiguous, then to be resolvable again (possibly to
+different bindings).
+
+Furthermore, there is some flexibility in the order in which macros can be
+expanded. How a name resolves should be consistent under any ordering.
+
+The strongest form of this principle, I believe, is that at any stage of
+macro expansion, and under any ordering of expansions, if a name resolves to a
+binding then it should always (i.e., at any other stage of any other expansion
+series) resolve to that binding, and if resolving a name produces an error
+(n.b., distinct from not being able to resolve), it should always produce an
+error.
+
+
+#### Avoid errors due to the resolver being stuck.
+
+Errors with concrete causes and explanations are easier for the user to
+understand and to correct. If an error is caused by name resolution getting
+stuck, rather than by a concrete problem, this is hard to explain or correct.
+
+For example, if we support a rule that means that a certain glob can't be
+expanded before a macro is, but the macro can only be named via that glob
+import, then there is an obvious resolution that can't be reached due to our
+ordering constraints.
+
+
+#### The order of declarations of items should be irrelevant.
+
+I.e., names should be able to be used before they are declared. Note that this
+clearly does not hold for declarations of variables in statements inside
+function bodies.
+
+
+#### Macros should be manually expandable.
+
+Compiling a program should have the same result before and after expanding a
+macro 'by hand', so long as hygiene is accounted for.
+
+
+#### Glob imports should be manually expandable.
+
+A programmer should be able to replace a glob import with a list import that
+imports any names imported by the glob and used in the current scope, without
+changing name resolution behaviour.
+
+
+#### Visibility should not affect name resolution.
+
+Clearly, visibility affects whether a name can be used or not. However, it
+should not affect the mechanics of name resolution. I.e., changing a name from
+public to private (or vice versa), should not cause more or fewer name
+resolution errors (it may of course cause more or fewer accessibility errors).
```

Comment:
I think the example you give should continue to compile - we totally bend these principles in several places and this seems like a good place where usability should trump principle. It does occur to me that one of the proposals below would introduce backwards incompatibility though:

```
struct S;

mod a {
    pub Struct S;
}

mod b {
    use super::*;
    use a::*;
}
```

I believe that compiles today, but would cause an ambiguity with this RFC.

Yeah, macro visibility is definitely for another RFC.